### PR TITLE
Speed up merlin generation

### PR DESCRIPTION
**Pull request by rgrinberg**
Generate the string all at once and reuse buffer. I'm not sure how much this helps performance since the old approach wasn't slow. But this piece of code does run on 0 rebuilds so it's worth tweaking. Also, this cleanup makes it easier to fix the order of paths following the symbolic path fix.

### CODE REVIEW

**Comment by diml** [comment]
Overall this seems fine to me, but if the motivation is performance the original code to compute the build and source directories looked more efficient to me as it avoided allocating many intermediate records.

**Comment by rgrinberg** [comment]
Oh okay, that can be easily avoided in my approach as well. I'll amend the
pr.

**Comment by rgrinberg** [comment]
@diml no more records.

Btw, I get a slight improvement on a large workspace with this.

File: src/merlin.ml [hunk 1]
```diff
@@ -31,6 +31,38 @@ module Preprocess = struct
...
+    Path.Set.iter obj_dirs ~f:(fun p ->
+      printf "B %s\n" (serialize_path p)
+    );
+    Path.Set.iter src_dirs ~f:(fun p ->
+      printf "S %s\n" (serialize_path p)
+    );
+    begin match ppx with
+    | [] -> ()
+    | ppx ->
+      printf "FLG -ppx %s\n" (Filename.quote (String.concat ~sep:" " ppx));
```
Comment (diml) [inline src/merlin.ml]: Before we were calling `quote_for_shell` for individual element and then concatenating the result, this is a bit different. I think we should be conservative here

**Pull request merged**